### PR TITLE
Pull checkstyle#14615: fix String template AST ordering, drop empty content nodes

### PR DESCRIPTION
**Pull request by nrmancuso**
The current AST building was buggy and unmaintainable, manually setting line numbers was crazy. Now that we have the context cache, we can readily break the string template tokens into smaller tokens to do much less work (reliably) when we build the AST.

 - As you can see in the diff, we were missing some tokens
 - Fixed token ordering
 - Improved AST building to do less manual code in the visitor

### CODE REVIEW

File: config/import-control.xml
Comment (nrmancuso) [inline config/import-control.xml:315]: Here

Comment (rnveach) [reply config/import-control.xml:315]: diff --git a/config/import-control.xml b/config/import-control.xml
index 4a4b613..222543b 100644
--- a/config/import-control.xml
+++ b/config/import-control.xml
@@ -309,8 +309,6 @@
     </file>
     <file name="CompositeLexerContextCache">
       <allow pkg="org.antlr.v4.runtime"/>
-    </file>
-    <file name="CompositeLexerContextCache">
       <allow class="com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageLexer"/>
     </file>
   </subpackage>

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java:1848]: This was crazy, glad to see this go.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4
Comment (nrmancuso) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:257]: I changed this to be a singular fragment and allow us to define the number of these fragments in the rule where this fragment is used.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4
Comment (nrmancuso) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:262]: Extra predicate here that will be helpful in the future when we start parsing text block templates.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4
Comment (nrmancuso) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:473]: Empty "content" is nonsense, producing this token for static analysis purposes is worthless.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4
Comment (nrmancuso) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4:794]: These parser rules are now more reflective of what's in the JLS: https://docs.oracle.com/javase/specs/jls/se21/preview/specs/string-templates-jls.html#:~:text=template%20(3.13).-,3.13%20Fragments,-A%20template%20(

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:528]: These needed to go.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:553]: As you can see, the old implementation caused us to have a bunch of out of order elements in the AST

**Review by nrmancuso — COMMENTED** [review]
Comments:

**Comment by nrmancuso** [comment]
@rnveach let's proceed with review on the final commit only so that we can get this merged and unblock https://github.com/checkstyle/checkstyle/pull/14390 ASAP

I will rebase this PR after we merge the first commit in https://github.com/checkstyle/checkstyle/pull/14568

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt
Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:528]: Related to https://github.com/checkstyle/checkstyle/pull/14615#discussion_r1518575615

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt:362]: I assume there is no easy way to prove we are not missing other text?
Note: There is no removal of `there`, so it seems this was completely gone from our AST.

**Comment by rnveach** [comment]
@nrmancuso Regression is good. But have we done any time regression for this PR? :)

**Comment by rnveach** [comment]
https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/rework-string-template/2024-03-09-T-16-30-11/antlr-report-checkstyle/checkstyle/index.html#A35
I am not seeing a removal of the `SINGLE_LINE_COMMENT` and only the addition. Does this mean your fix is finding other issues we don't know about? I don't remember seeing a test case for something like this.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt
Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt:362]: We could, it would be involved though. We could have some verification that all tokenized text exists in the AST.

1. Duplicate the token stream
2. Generate the tree (in code, not text, a 'la DetailAST)
3. Check to make sure that everything that is tokenized exists in the tree via n-nary search

The main reason that we were missing stuff here was due to the weird building of the AST that I used previously to try to keep the lexer as context free as possible.

**Comment by nrmancuso** [comment]
>https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/rework-string-template/2024-03-09-T-16-30-11/antlr-report-checkstyle/checkstyle/index.html#A35
I am not seeing a removal of the SINGLE_LINE_COMMENT and only the addition. I also do not see it mention above.

I will add a test like this to make sure we don't have any regressions.

>https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/rework-string-template/2024-03-09-T-16-30-11/antlr-report-checkstyle/checkstyle/index.html#A111
This is similarly a new item.

This one is captured in the AST tests, so we are covered there.

>Also what is your opinion of running antlr regression against a later openjdk project (is this 21 syntax?) which might have more of these type of code and more?

This is difficult, we usually OOM in Github actions on OpenJDK sadly. I have decommissioned my dedicated report generation box in favor of Github actions. I would like to find some time to make this work in CI somehow.

**Comment by rnveach** [comment]
OpenJDK 21 ANTLR Regression: https://rveach.no-ip.org/checkstyle/regression/342/

**Comment by romani** [comment]
This link is not working

**Comment by nrmancuso** [comment]
>https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/rework-string-template/2024-03-09-T-16-30-11/antlr-report-checkstyle/checkstyle/index.html#A35
I am not seeing a removal of the SINGLE_LINE_COMMENT and only the addition. I also do not see it mention above.

@rnveach test is added

File: src/test/java/com/puppycrawl/tools/checkstyle/grammar/java21/Java21AstRegressionTest.java
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammar/java21/Java21AstRegressionTest.java:59]: I've extended this test case to get our nesting level of curly braces even deeper within one "context", and left a comment here about this test case.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/InputStringTemplateNested.java
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/InputStringTemplateNested.java:17]: Previous test case only went to a nesting level of 1, we needed to go higher to kill a pitest mutation (and be more thorough in our testing in general).

**Comment by rnveach** [comment]
> This link is not working

IM me if you need it back @nrmancuso . I took it down since I thought we were done. It is up for now.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasicWithTabs.txt:235]: I don't see any changes to `TokenTypes` and our ANTLR displays. If we don't have a UT requiring these to be kept in synch, these kind of changes will break `TokenTypes`.

Also, did we remove all usages of `STRING_TEMPLATE_CONTENT `? Should this token now be undocumented?

**Review by rnveach — CHANGES_REQUESTED** [review]

**Review by rnveach — APPROVED** [review]

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:84]: we need to update https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#STRING_TEMPLATE_CONTENT 
only single token example.
we need one more compact example,  to show that STRING_TEMPLATE_CONTENT might not be present.

We are very inconsistent on presence and not presence.
example is `MODIFIERS` but there are more:
```
 VARIABLE_DEF -> VARIABLE_DEF
  |--MODIFIERS -> MODIFIERS
  |--TYPE -> TYPE
  |   `--IDENT -> String
  |--IDENT -> s
```

it is always tricky to deal with it.
I am not sure what is right way to make AST: 1) always define 2) skip if empty.
I am ok to skip. 
If it was not obvious for us, better to explicitly show it.

**Review by romani — CHANGES_REQUESTED** [review]
only one minor:

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt
Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:84]: My issue with it was that we call it "content" but it is an imaginary node that has no content, with the same line/column number that it's next sibling does. This is nonsense.

Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/ExpectedStringTemplateBasic.txt:84]: Example is updated

**Review by romani — APPROVED** [review]
ok to merge.

please update PR title to reference token that will be skipped.
PR/issue titile is used in release notes, users need to easily catch by keywords what changed.

fact of refactoring, user doesnot care.
They might be affected by missing token begining from this release.

**Comment by nrmancuso** [comment]
> ok to merge.
> 
> please update PR title to reference token that will be skipped. PR/issue titile is used in release notes, users need to easily catch by keywords what changed.
> 
> fact of refactoring, user doesnot care. They might be affected by missing token begining from this release.

done

**Pull request merged**
